### PR TITLE
feature-review

### PR DESCRIPTION
**Pull request by margierain**
merging changes to master

### CODE REVIEW

File: main/app.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+            self.people_data['Fellow'].append({name: want_accommodation})
+
+        self.insert_db(name=name, person_type=person_type,
+                       want_accommodation=want_accommodation)
+        self.allocation_rule(name, person_type,
+                             want_accommodation)
+
+    def allocation_rule(self, name, person_type, want_accommodation):
+        # to randomly allocate everyone an office
+        # ipdb.set_trace()
```
Comment (afropolymath) [inline main/app.py]: **Code Quality: Nitpick**

Remove all commented out code in your codebase.

File: main/app.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+            'L': {' ': []}
+        }
+        # variable used in allocation
+        self.offices = []
+        self.full_name = []
+        self.allocated_room = []
+
+        # variables used in add_person
+        self.name = ""
+        name = self.name
```
Comment (afropolymath) [inline main/app.py]: **Code Qulaity: Nitpick**
You could make use of multiple assignments here to shorten code. Also on line 44/45.

``` py
name = self.name = ""
```

File: models/personel.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+class Personel(object):
+    """personel model"""
+
+    def __init__(self, name):
+        self.name = name
+        want_accommodation = 'N'
+
+
+class staff(Personel):
```
Comment (afropolymath) [inline models/personel.py]: **Code Quality: Critical**

Always use upper case for your class names. See PEP-8.

File: models/personel.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+    def __init__(self, name):
```
Comment (afropolymath) [inline models/personel.py]: **Code Quality: Nitpick**

You should take advantage of the superclass constructor to ensure variables like `wants_accomodation` are actually set on object creation.

File: models/personel.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
+class Personel(object):
+    """personel model"""
+
+    def __init__(self, name):
+        self.name = name
+        want_accommodation = 'N'
```
Comment (afropolymath) [inline models/personel.py]: **Code Quality: Nitpick**

I presume this is an instance variable. You should use `self.wants_accommodation` if that's the case.

File: models/room.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+        pass
+
+
+class Office(Room):
+    """office model has capacity of 6"""
+    room_type = 'O'
+    capacity = 6
+
+
+class Living_space(Room):
```
Comment (afropolymath) [inline models/room.py]: **Code Quality: Critical**

Use `PascalCase` for class names. This should be `LivingSpace` and not `Living_space`. See PEP-8.

File: models/room.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
+
+class Room(object):
+    """Room is the parent and Office and Living_space inherit from it"""
+
+    def __init__(self):
+        """both Office and Living_space have access to name"""
```
Comment (afropolymath) [inline models/room.py]: **Code Quality: Nitpick**

Shouldn't rooms have names? If so where are they being stored?

File: main/app.py [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+        self.rooms = {
+            'O': [],
+            'L': []
+        }
+        # variable used in add_person
+        self.people_data = {'Staff': [],
+                            'Fellow': []
+                            }
+        self.allocate = {
+            'O': {' ': []},
```
Comment (afropolymath) [inline main/app.py]: **Code Quality: Critical**

You just created a dictionary with an empty string as a key. No matter why you did this, it's definitely not good practice.

File: main/app.py [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+            'L': {' ': []}
+        }
+
+        self.room_type = "O" or "L"
+        self.name = ""
+        name = self.name
+        self.personnel_type = ""
+        personnel_type = self.personnel_type
```
Comment (afropolymath) [inline main/app.py]: **Code Quality: Nitpick**

I thought we agreed on one-line multiple assignments for this.

File: main/app.py [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+        ''' keep track of offices allocated'''
+        off_name = str(off_name)
+        office_space = self.connect.execute(
+            "SELECT COUNT(*) AS office_occupants FROM Persons  WHERE Persons.office_accommodation = ?", [off_name]).fetchall()
+        self.office_occupied = office_space[0][0]
+        return self.office_occupied
+
+    def living_space_count(self, liv_name):
+        '''keep track of living space allocated'''
+        # print(liv_name)
```
Comment (afropolymath) [inline main/app.py]: **Code Quality: Nitpick**

Make it a habit to rid your code of all the commented out code before submitting.

File: main/app.py [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+                self.rooms[room_type.upper()].append(room_name)
+            else:
+                self.rooms['L'].append(room_name)
+            self.connect.execute(
+                "INSERT INTO Rooms (Name, type) VALUES (?, ?)", [room_name, room_type])
+            self.conn.commit()
+        print('New rooms succesfully created')
+
+    def office_space_count(self, off_name):
+        ''' keep track of offices allocated'''
```
Comment (afropolymath) [inline main/app.py]: **Code Quality: Nitpick**

You should use proper capitalization for Doc string comments like these.

File: main/models/room.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
+
+class Room(object):
+    """Room is the parent and Office and Living_space inherit from it"""
+
+    def __init__(self, connection):
+        self.connect = connection
+        #self.name = name
```
Comment (afropolymath) [inline main/models/room.py]: **Code Quality: Nitpick**

Remove this line. Also not sure what `connection` is doing here.

File: tests/room_test.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
+import unittest
+from ..main.app import Amity
+import sqlite3
+
+import os
+
+
+
+class roomstest(unittest.TestCase):
```
Comment (afropolymath) [inline tests/room_test.py]: **Testing: Nitpick**

You did not write enough tests here**

File: tests/room_test.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+    def setUp(self):
+        self.amity = Amity()
+        self.conn = sqlite3.connect("test_amity.sqlite")
+        self.connect = self.conn.cursor()
+        self.connect.execute(
```
Comment (afropolymath) [inline tests/room_test.py]: **Testing: Nitpick**

You should not be running SQL logic of this sort directly within your tests. You should call a function that does this.

File: tests/room_test.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+                            'Fellow': []
+                            }
+
+
+    def test_add_person(self):
+        self.amity.add_person('Margie', 'Rain', 'Fellow', 'Y')
+        self.name = 'Margie' + " " + 'Rain'
+        self.person_type = 'Fellow'
+        self.want_accommodation = 'Y'
+        self.assertEqual(self.name, 'Margie Rain')
```
Comment (afropolymath) [inline tests/room_test.py]: **Code Quality: Critical**

Why are you testing a value that you just set 3 lines above? Need to work on having a better understanding of Testing.

File: main/app.py (hunk 8 above)
Comment (margierain) [inline main/app.py]: corrected

File: load_data.txt [hunk 16]
```diff
@@ -0,0 +1,6 @@
+HANNAH KOSKE FELLOW Y
+BENJAMIN KAGIA FELLOW Y
```
Comment (afropolymath) [inline load_data.txt:2]: **Code Quality: Nitpick**

You should use a larger test file to enable you check for more edge cases

File: main/app.py [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+
+        self.living_occupied = 0
+        self.office_occupied = 0
+        file = self.file = " "
+        allocated = self.allocated = []
+
+        self.conn = sqlite3.connect("amity.sqlite")
+        self.conn.row_factory = sqlite3.Row
+        self.connect = self.conn.cursor()
+        self.connect.execute(
```
Comment (afropolymath) [inline main/app.py]: **Code Quality: Nitpick**

Always run your code within a `try...except` block to take care of error cases.

File: main/app.py [hunk 18]
```diff
@@ -0,0 +1,352 @@
...
+    def create_rooms(self, room):
+        """Allows user to enter a list of room names specifying
+                whether office or living spaces"""
+        room_type = raw_input(
+            "Enter room type: \n O: Office space \n L: Living space: \n")
+
+        while room_type != "O" and room_type != "L":
+            room_type = raw_input(
+                "Try again. Enter Room Type:\n O: Office space \n L: Living space: \n")
+        for room_name in room:
```
Comment (afropolymath) [inline main/app.py:96]: **Code Quality: Nitpick**

Not sure where this `room_name` variable is coming from.

File: main/app.py [hunk 19]
```diff
@@ -0,0 +1,352 @@
...
+                self.rooms[room_type.upper()].append(room_name)
+            else:
+                self.rooms['L'].append(room_name)
+            self.connect.execute(
+                "INSERT INTO Rooms (Name, type) VALUES (?, ?)", [room_name, room_type])
+            self.conn.commit()
+        print('New rooms succesfully created')
+        return room
+
+    def office_space_count(self, off_name):
```
Comment (afropolymath) [inline main/app.py]: **Code Quality: Nitpick**

This should be a property of the office class. Following OO design, the office object should be the only entity with that information.

File: main/app.py [hunk 20]
```diff
@@ -0,0 +1,352 @@
...
+        '''Keep track of offices allocated should be less than 6'''
+        off_name = str(off_name)
+        office_space = self.connect.execute(
+            "SELECT COUNT(*) AS office_occupants FROM Persons  WHERE Persons.office_accommodation = ?", [off_name]).fetchall()
+        self.office_occupied = office_space[0][0]
+        return self.office_occupied
+
+    def living_space_count(self, liv_name):
+        '''Keep track of living space allocated should be less than 4'''
+        # print(liv_name)
```
Comment (afropolymath) [inline main/app.py]: **Code Quality: Nitpick**

Remember to clean out your codebase always. Don't leave commented out code in here.

File: main/app.py [hunk 21]
```diff
@@ -0,0 +1,352 @@
...
+            print(self.living_occupied)
+        return self.living_occupied
+
+    def add_person(self, first_name, last_name, person_type, want_housing):
+        '''Method receives personnel data and calls other method to save and allocate rooms to personnel '''
+        name = first_name + " " + last_name
+        person_type = person_type
+        want_accommodation = want_housing
+        if person_type.upper() == 'STAFF':
+            self.people_data['Staff'].append(name)
```
Comment (afropolymath) [inline main/app.py:112]: **Code Quality: Nitpick**

I would prefer it if your data structures were somewhat consistent. It appears the two lists in question here are housing two different types of objects. Might be a better approach to use the same sub-objects to ensure parsing the data later on is much easier.

File: main/app.py [hunk 22]
```diff
@@ -0,0 +1,352 @@
...
+        else:
+            self.people_data['Fellow'].append({name: want_accommodation})
+
+        self.insert_db(name=name, person_type=person_type,
```
Comment (afropolymath) [inline main/app.py:116]: **Code Quality: Nitpick**

What happens here in the case where we have a staff. I'm referring to the `want_accomodation` variable.

File: main/app.py [hunk 23]
```diff
@@ -0,0 +1,352 @@
...
+                       want_accommodation=want_accommodation)
+
+        self.allocation_rule(name, person_type,
```
Comment (afropolymath) [inline main/app.py:119]: **Code Quality: Nitpick**

Also not sure what happens with the `want_accomodation` variable. Please reply with an explanation.

File: main/app.py [hunk 24]
```diff
@@ -0,0 +1,352 @@
...
+        return name +" " + person_type + " " + want_accommodation
+
+    def get_rooms(self, r_type):
+        return self.connect.execute(
+            "SELECT Name FROM Rooms where type = '" + r_type + "'").fetchall()
+
+    def allocation_rule(self, name, person_type, want_accommodation):
+        '''Randomly allocate everyone an office'''
+        office_rooms = self.get_rooms('O')
+        for office_room in office_rooms:
```
Comment (afropolymath) [inline main/app.py]: **Code Quality: Critical**

This is extremely sub-optimal. You run this loop every time to get available offices. Can we have this information persistent somewhere and then remove an office/room as the case maybe when the room has been filled. That's a much better approach.

File: main/app.py [hunk 25]
```diff
@@ -0,0 +1,352 @@
...
+
+        self.allocation_rule(name, person_type,
+                             want_accommodation)
+        return name +" " + person_type + " " + want_accommodation
+
+    def get_rooms(self, r_type):
+        return self.connect.execute(
+            "SELECT Name FROM Rooms where type = '" + r_type + "'").fetchall()
+
+    def allocation_rule(self, name, person_type, want_accommodation):
```
Comment (afropolymath) [inline main/app.py:126]: **Code Quality: Nitpick**

Not sure how this function name defines what this function is doing.

File: main/app.py [hunk 26]
```diff
@@ -0,0 +1,352 @@
...
+
+            if self.housing:
+                self.allocate_housing(name, self.housing)
+            else:
+                print("No living space available")
+        else:
+            print("Personnel hasn't requested to be housed")
+
+
+    def insert_db(self, **kwargs):
```
Comment (afropolymath) [inline main/app.py:166]: **Code Quality: Nitpick**

I feel like you will be able to do a proper implementation of `save_state` if this function were omitted.

File: main/app.py [hunk 27]
```diff
@@ -0,0 +1,352 @@
...
+        for office_room in office_rooms:
+            room = Office(self.connect)
+            office_name = office_room[0]
+
+            if not room.is_room_filled(office_name):
+                self.available_office.append(office_name)
+
+        self.office = random.choice(self.available_office)
+        if self.office:
+            self.allocate_office(name, self.office)
```
Comment (afropolymath) [inline main/app.py]: **Code Quality: Critical**

What happens if an error occurs here? Code doesn't seem robust enough.

File: main/app.py (hunk 22 above)
Comment (margierain) [inline main/app.py:116]: by default staff aren't to accommodated so it is set to 'N'  in run.py

File: main/app.py (hunk 23 above)
Comment (margierain) [inline main/app.py:119]: this is taken care of in run.py file...

```
@parser_cmd
    def do_add_person(self, arg):
        """Usage: add_person <first_name> <last_name> <person_type>
        [--wa=n]"""
        if type(arg['<first_name>']) != str or type(arg['<last_name>']) != str or type(arg['<person_type>']) != str:
            raise ValueError

        arg['--wa'] = 'Y' if arg['--wa'] is not None and arg['--wa'].upper() == 'Y' else 'N'
        amity.add_person(arg['<first_name>'], arg['<last_name>'], arg['<person_type>'], arg['--wa'])
```

File: main/app.py (hunk 26 above)
Comment (margierain) [inline main/app.py:166]: Do you mean I should change the name of the function
